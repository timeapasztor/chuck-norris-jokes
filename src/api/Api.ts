export default class Api {
    public static fetchRandomJoke = async () => {
        return await Api?.fetch(Url.randomJoke);
    };

    public static fetchJokeCategories = async () => {
        return await Api?.fetch(Url.jokeCategories);
    };

    public static fetchJokeFromCategory = async (category: string) => {
        return await Api?.fetch(`${Url.jokeFromCategory}${category}`);
    };

    public static fetchSearchResult = async (query: string) => {
        return await Api?.fetch(`${Url.searchQuery}${query}`);
    };

    private static fetch = async (url: string) => {
        try {
            let response = await fetch(url);
            return await response.json();
        } catch (exception) {
            return exception;
        }
    };
}

export const Url = {
    randomJoke: 'https://api.chucknorris.io/jokes/random',
    jokeCategories: 'https://api.chucknorris.io/jokes/categories',
    jokeFromCategory: 'https://api.chucknorris.io/jokes/random?category=',
    searchQuery: 'https://api.chucknorris.io/jokes/search?query='
};
