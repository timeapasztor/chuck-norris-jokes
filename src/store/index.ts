import { combineReducers, configureStore } from '@reduxjs/toolkit';
import { selectedCategoryReducer } from '../reducer/SelectedCategory';
import { randomJokeReducer } from '../reducer/RandomJoke';

export const store = configureStore({
    reducer: {
        selected: selectedCategoryReducer,
        randomJoke: randomJokeReducer
    }
});

export const rootReducer = combineReducers({
    selected: selectedCategoryReducer,
    randomJoke: randomJokeReducer
});

export type RootState = ReturnType<typeof rootReducer>;
