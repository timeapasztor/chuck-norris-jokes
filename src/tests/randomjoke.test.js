import React, { useEffect } from 'react';
import { create, act } from 'react-test-renderer';
import background from '../style/images/background.png';
import Api from '../api/Api';
import { useDispatch, useSelector } from 'react-redux';
import allActions from '../actions';
import { store } from '../store';
import { Provider } from 'react-redux';
import { Button } from '@material-ui/core';

function JokeDescription(props) {
    return <div>{props.joke}</div>;
}

function JokeCardImage(props) {
    return <img src={props.src} alt={props.alt} />;
}

const RandomJoke = () => {
    const randomJoke = useSelector((state) => {
        return state.randomJoke;
    });
    const dispatch = useDispatch();

    useEffect(() => {
        loadRandomJoke();
    }, []);

    const loadRandomJoke = async () => {
        try {
            let result = await Api.fetchRandomJoke();
            dispatch(allActions.RandomJoke.randomJoke(result));
        } catch (error) {
            console.log('error while fetching a random joke');
        }
    };

    return (
        <>
            <JokeDescription>{randomJoke?.value}</JokeDescription>
            <Button
                onClick={() => {
                    loadRandomJoke();
                }}
            >
                Get me a new one ...
            </Button>
        </>
    );
};

describe('Random joke description', () => {
    test('Matches the snapshot', () => {
        const div = create(<JokeDescription joke="Chuck Norris joke" />);
        expect(div.toJSON()).toMatchSnapshot();
    });
});

describe('Image on random joke page', () => {
    test('Matches the snapshot', () => {
        const img = create(<JokeCardImage src={background} alt="signature of chuck norris" />);
        expect(img.toJSON()).toMatchSnapshot();
    });
});

it('First load of a random joke', () => {
    let tree;
    act(() => {
        tree = create(
            <Provider store={store}>
                <RandomJoke />
            </Provider>
        );
    });
    expect(tree.toJSON()).toMatchSnapshot();
});

test('the className of Button includes MuiButton-root', () => {
    const root = create(<Button>Get me a new one ...</Button>).root;
    const element = root.findByType('button');
    expect(element.props.className.includes('MuiButton-root')).toBe(true);
});
