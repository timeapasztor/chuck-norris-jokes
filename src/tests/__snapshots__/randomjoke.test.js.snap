// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Matches the snapshot 1`] = `
<button>
  Nothing to do for now
</button>
`;

exports[`Clicking on button Get me a new one 1`] = `
Array [
  <div />,
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-text"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Get me a new one ...
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>,
]
`;

exports[`First load of a random joke 1`] = `
Array [
  <div />,
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-text"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Get me a new one ...
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>,
]
`;

exports[`Image on random joke page Matches the snapshot 1`] = `
<img
  alt="signature of chuck norris"
  src="background.png"
/>
`;

exports[`Loading random jokes 1`] = `<div />`;

exports[`Random joke description Matches the snapshot 1`] = `
<div>
  Chuck Norris joke
</div>
`;

exports[`it works 1`] = `
<div
  className="sc-bczRLJ insazo"
/>
`;

exports[`your test name 1`] = `<div />`;
